File: 3217-delete-nodes-from-linked-list-present-in-array.py
from typing import List, Optional

# Definition for singly-linked list.


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def modifiedList(self, nums: List[int], head: Optional[ListNode]) -> Optional[ListNode]:
        '''
        We can convert nums into set and go through head and if a node does not exist in the set, update the prev link to the next link. We will need to keep track of a prev node
        '''
        prev = ListNode(-1, head)
        origin = prev
        cur = head
        remove = set(nums)
        while cur:
            if cur.val in remove:
                prev.next = cur.next
            else:
                prev = cur
            cur = cur.next

        return origin.next


sol = Solution()
print(
    f'output: {sol.modifiedList([1, 2, 3], ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5))))))}')
print(
    f'output: {sol.modifiedList([1], ListNode(1, ListNode(2, ListNode(1, ListNode(2, ListNode(1, ListNode(2)))))))}')

File: 3217-delete-nodes-from-linked-list-present-in-array.java
import java.util.*;

class DeleteNodesFromLinkedListPresentInArray {
    /**
     * Definition for singly-linked list.
     */
    public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode modifiedList(int[] nums, ListNode head) {
        ListNode dummy = new ListNode(-1, head);
        ListNode prev = dummy;
        ListNode cur = head;
        Set<Integer> seen = new HashSet<>();
        for (int num : nums) {
            seen.add(num);
        }

        while (cur != null) {
            if (seen.contains(cur.val)) {
                prev.next = cur.next;
            } else {
                prev = cur;
            }
            cur = cur.next;
        }

        return dummy.next;
    }

    public static void main(String[] args) {

        DeleteNodesFromLinkedListPresentInArray sol = new DeleteNodesFromLinkedListPresentInArray();
        // To access the inner class ListNode, create it via the outer class instance
        ListNode testHead = sol.new ListNode(1, 
            sol.new ListNode(2, 
                sol.new ListNode(3, 
                    sol.new ListNode(4, 
                        sol.new ListNode(5)))));
        System.out.println("Output: " + sol.modifiedList(new int[] { 1, 2, 3 }, testHead));
    }
}
